text/0000-min-rust-version.md

```diff
@@ -0,0 +1,179 @@
+functionality behind it. The reason for it is to reduce implementation cost of
+the first stage to the minimum and ideally ship it as part of Rust 2018.
+It will also allow crate authors who care about MSRV to start mark their crates
+early.
+
+## Second stage: versions resolution
+
+`rust` field becomes required and cargo will add it as a constraint to dependency
+versions resolution. If user uses e.g. Rust 1.40 and uses crate `foo = "0.2"`, but
+all selected versions of `foo` specify MSRV e.g. equal 1.41 or bigger (or even
+nightly) `cargo` will issue an error.
+
+`rust` field value will be checked as well, on crate build `cargo` will check if
+all upstream dependencies can be built with the specified MSRV. (i.e. it will
+check if there is exists solution for given crates and Rust versions constraints)
+
+Yanked crates will be ignored in this process.
+
+Implementing this functionality hopefully will allow to close the debate regarding
+MSRV handling in crate versions and will allow crate authors to feel less
+restrictive about bumping their crate's MSRV. (though it can be a usefull
+convention for post-1.0 crates to bump minor version on MSRV change to allow
+publishing backports which fix serious issues using patch version)
+
+## Third stage: better crate checks
+
+Here we introduce two-level check for crates. First level will check if all used
+items were stabilised before or on given MSRV using `#[stable(since=version)]`
+attribute, issuing compile errors otherwise.
+
+Second level will try to build crate with the specified MSRV on `cargo publish`,
+i.e. words it will be required to install MSRV toolchain. (this check can be
+disabled using `--no-verify` option)
+
+While these two checks will not replace proper CI testing, they will help to
+reduce number of improper MSRV configuration to the minimum.
+
+Note that `rust` field must be equal to MSRV with default features for all
+supported targets.
+
+## Extension: nightly versions
+
+For some bleeding-edge crates which experience frequent breaks on Nightly updates
+(e.g. `rocket`) it can be useful to specify exact Nigthly version(s) on which
+crate can be built. One way to achieve this is by using the following syntax:
+- single version: rust = "nightly: 2018-01-01"
+- enumeration: "nightly: 2018-01-01, 2018-01-15"
+- (inclusive) range: "nightly: 2018-01-01..2018-01-15"
+- enumeration+range: "nightly: 2018-01-01, 2018-01-08..2018-01-15"
+
+Such restrictions can be quite severe, but hopefully this functionality will be
+used only by handful of crates.
+
+## Extension: cfg based MSRV
+
+Some crates can have different MSRVs depending on target architecture or enabled
+features. In such cases it can be usefull to extend `rust` field, e.g. in the
+following way:
+```toml
+rust = "1.30"
+rust-cases = [
+    { cfg = "x86_64-pc-windows-gnu", version = "1.35" },
+    { cfg = 'cfg(feature = "foo")', version = "1.33" },
+]
```

Comment:
I prefer we reuse the existing [`[target.*]` section](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#platform-specific-dependencies) for this:

```toml
rust = "1.30"

[target.x86_64-pc-windows-gnu]
rust = "1.35"

[target.'cfg(feature = "foo")']
rust = "1.33"
```

(Note that cargo should either fully support `[target.'cfg(feature = "foo")'.dependencies]` or outright reject it)

Reply:
Hm, reusing `target` is a good idea. I'll replace `rust-case` with your proposal.

As for `[target.'cfg(feature = "foo")'.dependencies]`, while it's somewhat redundant to `optional = true` dependencies + `foo = [".."]`, I think it could provide a better flexibility in specification of optional dependencies (e.g. add dependency only for given target if feature is enabled), so in my opinion it should be supported.